### DISCUSSION TITLE
Clarifying behavior with `graphql-codegen`

### DISCUSSION THREAD

**Discussion by aryzing** [Urigo/graphql-scalars#1407] [Q&A]
There are a few issues and discussions on how to use this package together with `@graphql-codegen/cli` ([docs](https://www.graphql-code-generator.com/docs/getting-started), [repo](https://github.com/dotansimha/graphql-code-generator)).

* Discussion https://github.com/Urigo/graphql-scalars/discussions/1181
* Discussion https://github.com/Urigo/graphql-scalars/discussions/1215
* Issue https://github.com/Urigo/graphql-scalars/issues/953
* Issue https://github.com/Urigo/graphql-scalars/issues/419
* PR https://github.com/dotansimha/graphql-code-generator/pull/2418
* PR https://github.com/dotansimha/graphql-code-generator/pull/2600

Together, these issues and discussions are fundamentally trying to achieve the same objective: typed scalars.

Given the flexibility of this ecosystem, there's more than one way to do this. However, it seems that there are no concrete examples or docs for popular workflows. Specifically,

**1. How to use the `codegenScalarType` to generate typed scalars?**

There seems to be no concrete, step-by-step examples on how to do this.

**2. When using SDL, how to link `graphql-scalars`' types into `@graphql-codegen/cli`**

When using SDL, seems that the only way to obtain typed scalars with `@graphql-codegen/cli` is to manually set their type in `.graphqlrc.yaml`,

```yaml
schema: schema.graphql
extensions:
  codegen:
    generates:
      src/generated/types.ts:
        plugins:
          - typescript
        config:
          scalars:
            DateTime: "Date | string"
```

Furthermore, it feels weird that an executable schema can be created by combining project-specific definitions and library definitions, but can't be done for code generation. Let me illustrate,

```ts
// Combine project type defs with `graphql-scalars` type defs
// typeDefs.ts
import { readFileSync } from "fs";
import { DateTimeTypeDefinition } from "graphql-scalars";
const projectTypeDefs = fs.readFileSync("schema.graphql").toString()
export const typeDefs = [DateTimeTypeDefinition, projectTypeDefs];

// Combine project resolvers with `grapql-scalars` resolvers
// resolvers.ts
import { DateTimeResolver } from "graphql-scalars";
export const resolvers = {
  DateTime: DateTimeResolver,
  Query: {/**/},
  Mutation: {/**/},
};

// Make executable schema
// server.ts
import { makeExecutableSchema } from "@graphql-tools/schema";
import { resolvers } from "resolvers";
import { typeDefs } from "typeDefs";

const schema = makeExecutableSchema({
  typeDefs,
  resolvers,
});
```

However, when using `@graphql-codegen/cli`, *how can `DateTime` be injected into the generation process*?

The best I've been able to come up with, is to re-declare `scalar DateTime` in the project's own `schema.graphql`, although this is technically redundant. But it works. 

Regardles, best practices and guidance is what I'm after. Is this 👆 the best way to include scalars into the code generation process?

**Comment by dchambers** [comment]
Not sure it will be exactly what you want, but this is how I resolved it in my case:

- https://github.com/dotansimha/graphql-code-generator/pull/2418#issuecomment-1052727007
